### PR TITLE
Upstream many charon queries

### PR DESCRIPTION
**Pull request by Nadrieril**
The main change of this PR is the addition of the `hax::FullDefKind` enum, which reflects rustc's `DefKind` but with a lot of extra information. This will be the entrypoint for charon's translations; I've already been able [to noticeably simplify charon](https://github.com/AeneasVerif/charon/pull/296) by using this PR.

### CODE REVIEW

**Comment by Nadrieril** [comment]
Marking it as draft while I debug why I'm handling closures wrong

**Comment by Nadrieril** [comment]
[Companion PR](https://github.com/AeneasVerif/charon/pull/296) passes CI this can be merged!

**Comment by W95Psp** [comment]
I just looked at the PR quickly, from what I understand FullDefKind doesn't ship any bodies, right?
I.e. shouldn't `FullDefKind::Fn` provide a body?

**Comment by Nadrieril** [comment]
Yeah, I haven't figured out how I want to do bodies yet. There are other things missing anyway (e.g. related to trait solving), I was thinking this PR is big enough already

**Comment by W95Psp** [comment]
Ok, that makes sense, let's postpone that to another PR :)
Best thing there would be to use a generic Body so that FullDefKind works for both MIR and THIR

File: frontend/exporter/src/types/copied.rs [hunk 1]
```diff
@@ -67,6 +77,380 @@ impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, GlobalIdent> for rustc_hir::def_id
...
+    /// - `bar` in `struct Foo { bar: u8 }`
+    /// - `Foo::Bar::0` in `enum Foo { Bar(u8) }`
+    Field,
+    /// Lifetime parameter: the `'a` in `struct Foo<'a> { ... }`
+    LifetimeParam,
+    /// A use of `global_asm!`.
+    GlobalAsm,
+}
+
+impl FullDef {
```
Comment (W95Psp) [inline frontend/exporter/src/types/copied.rs:382]: Seeing how lengthy things are, I wonder if we should not move `FullDefKind` and impls on it to its own file in `types/new`.

File: frontend/exporter/src/types/copied.rs [hunk 2]
```diff
@@ -67,6 +77,380 @@ impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, GlobalIdent> for rustc_hir::def_id
...
+#[derive(Clone, Debug, JsonSchema)]
+pub struct FullDef {
+    #[value(self.sinto(s))]
+    pub def_id: DefId,
+    #[value(s.base().tcx.opt_parent(*self).sinto(s))]
+    pub parent: Option<DefId>,
+    #[value(s.base().tcx.def_span(*self).sinto(s))]
+    pub span: Span,
+    #[value(s.base().tcx.def_key(*self).disambiguated_data.sinto(s))]
+    pub path_item: DisambiguatedDefPathItem,
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply frontend/exporter/src/types/copied.rs:162]: Oh, I did not know that your `DefId` included that. Indeed that's redundant.

File: frontend/exporter/src/types/copied.rs (hunk 1 above)
Comment (Nadrieril) [reply frontend/exporter/src/types/copied.rs:382]: Good idea

File: frontend/exporter/src/types/copied.rs [hunk 3]
```diff
@@ -67,6 +77,380 @@ impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, GlobalIdent> for rustc_hir::def_id
...
+    let predicates = s.base().tcx.predicates_defined_on(def_id);
+    let pred_list = process_generic_predicates(s, predicates.predicates);
+    GenericPredicates {
+        parent: predicates.parent.sinto(s),
+        predicates: pred_list,
+    }
+}
+
+#[cfg(feature = "rustc")]
+fn get_item_predicates<'tcx, S: UnderOwnerState<'tcx>>(s: &S, def_id: RDefId) -> GenericPredicates {
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply frontend/exporter/src/types/copied.rs:126]: They both call `process_generic_predicates` which is a bit of a hack so I'm not in favor of moving them to an impl block. I added docs.

[the remark this answers is not in the document]
Comment (W95Psp) [reply frontend/exporter/src/types/copied.rs]: Ok, makes sense!

File: frontend/exporter/src/types/copied.rs [hunk 4]
```diff
@@ -67,6 +77,380 @@ impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, GlobalIdent> for rustc_hir::def_id
     }
 }
 
+#[cfg(feature = "rustc")]
+fn process_generic_predicates<'tcx, S: UnderOwnerState<'tcx>>(
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply frontend/exporter/src/types/copied.rs]: Done

File: frontend/exporter/src/types/copied.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nadrieril) [reply frontend/exporter/src/types/copied.rs]: I simplified it a bit more

**Pull request merged**